spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteTablePathSparkAction.java

```diff
@@ -281,24 +281,34 @@ private String rebuildMetadata() {
     // rebuild version files
     RewriteResult<Snapshot> rewriteVersionResult = rewriteVersionFiles(endMetadata);
     Set<Snapshot> deltaSnapshots = deltaSnapshots(startMetadata, rewriteVersionResult.toRewrite());
-
-    Set<String> manifestsToRewrite = manifestsToRewrite(deltaSnapshots, startMetadata);
     Set<Snapshot> validSnapshots =
         Sets.difference(snapshotSet(endMetadata), snapshotSet(startMetadata));
 
+    // rebuild manifest files
+    Set<ManifestFile> manifestsToRewrite = manifestsToRewrite(validSnapshots);
+
+    Map<String, RewriteContentFileResult> rewriteManifestResult =
+        rewriteManifests(deltaSnapshots, endMetadata, manifestsToRewrite);
+
+    // Extract manifest file sizes for manifest list rewriting
+    Map<String, Long> rewrittenManifestLengths =
+        rewriteManifestResult.entrySet().stream()
+            .collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().length()));
+
     // rebuild manifest-list files
     RewriteResult<ManifestFile> rewriteManifestListResult =
         validSnapshots.stream()
-            .map(snapshot -> rewriteManifestList(snapshot, endMetadata, manifestsToRewrite))
+            .map(snapshot -> rewriteManifestList(snapshot, endMetadata, rewrittenManifestLengths))
             .reduce(new RewriteResult<>(), RewriteResult::append);
 
-    // rebuild manifest files
-    RewriteContentFileResult rewriteManifestResult =
-        rewriteManifests(deltaSnapshots, endMetadata, rewriteManifestListResult.toRewrite());
+    // Aggregate all manifest rewrite results
+    RewriteContentFileResult allManifestsResult =
+        rewriteManifestResult.values().stream()
+            .reduce(new RewriteContentFileResult(), RewriteContentFileResult::append);
 
     // rebuild position delete files
     Set<DeleteFile> deleteFiles =
-        rewriteManifestResult.toRewrite().stream()
+        allManifestsResult.toRewrite().stream()
```

Comment:
nit: is the variable renaming more accurate? trying to see if it is unnecessary change.